File: Pandacap.LowLevel/Data/TheirContent/ActivityPub/ActivityPubFavorite.fs
namespace Pandacap.Data

open System
open System.ComponentModel.DataAnnotations
open Pandacap.ActivityPub
open Pandacap.Html
open Pandacap.PlatformBadges

/// An image attachment to a remote ActivityPub post that this app's instance owner has liked.
type ActivityPubFavoriteImage() =
    member val Url = "" with get, set
    member val Name = nullString with get, set

/// A remote ActivityPub post that this app's instance owner has liked.
type ActivityPubFavorite() =
    [<Key>]
    member val Id = Guid.Empty with get, set

    member val ObjectId = "" with get, set
    member val CreatedBy = "" with get, set
    member val Username = nullString with get, set
    member val Usericon = nullString with get, set
    member val CreatedAt = DateTimeOffset.MinValue with get, set
    member val FavoritedAt = DateTimeOffset.MinValue with get, set
    member val HiddenAt = nullDateTimeOffset with get, set
    member val Summary = nullString with get, set
    member val Sensitive = false with get, set
    member val Name = nullString with get, set
    member val Content = nullString with get, set
    member val InReplyTo = nullString with get, set
    member val Attachments = new ResizeArray<ActivityPubFavoriteImage>() with get, set

    interface IActivityPubLike with
        member this.ObjectId = this.ObjectId

    interface IFavorite with
        member this.HiddenAt
            with get () = this.HiddenAt
             and set value = this.HiddenAt <- value

        member this.FavoritedAt = this.FavoritedAt

    interface IPost with
        member _.Platform = ActivityPub
        member this.Url = this.ObjectId
        member this.DisplayTitle =
            if not (String.IsNullOrWhiteSpace(this.Name)) then
                this.Name
            else
                this.Content
                |> TextConverter.FromHtml
                |> TitleGenerator.FromBody
        member this.Id = $"{this.Id}"
        member this.InternalUrl = $"/RemotePosts?id={Uri.EscapeDataString(this.ObjectId)}"
        member this.ExternalUrl = $"{this.ObjectId}"
        member this.PostedAt = this.CreatedAt
        member this.ProfileUrl = this.CreatedBy
        member this.Thumbnails = [
            if not this.Sensitive then
                for a in this.Attachments do
                    if not (String.IsNullOrEmpty(a.Url)) then {
                        new IPostThumbnail with
                            member _.AltText = a.Name
                            member _.Url = a.Url
                    }
        ]
        member this.Usericon = this.Usericon
        member this.Username = this.Username

File: Pandacap.Database/ActivityPubFavorite.cs
using Pandacap.ActivityPub.Models.Interfaces;
using Pandacap.Text;
using Pandacap.UI.Badges;
using Pandacap.UI.Elements;

namespace Pandacap.Database
{
    public class ActivityPubFavorite : IFavorite, IActivityPubLike
    {
        public Guid Id { get; set; }

        public string ObjectId { get; set; } = "";

        public string CreatedBy { get; set; } = "";

        public string? Username { get; set; }

        public string? Usericon { get; set; }

        public DateTimeOffset CreatedAt { get; set; }

        public DateTimeOffset FavoritedAt { get; set; }

        public DateTimeOffset? HiddenAt { get; set; }

        public string? Summary { get; set; }

        public bool Sensitive { get; set; }

        public string? Name { get; set; }

        public string? Content { get; set; }

        public string? InReplyTo { get; set; }

        public class Image : IPostThumbnail
        {
            public string Url { get; set; } = "";

            public string? Name { get; set; }

            string IPostThumbnail.AltText => Name ?? "";
        }

        public List<Image> Attachments { get; set; } = [];

        Badge IPost.Badge => Badges.ActivityPub.WithHostFromUriString(ObjectId);

        string IPost.DisplayTitle => Name ?? TextConverter.FromHtml(Content);

        string IPost.Id => $"{Id}";

        string IPost.InternalUrl => $"/RemotePosts?id={Uri.EscapeDataString(ObjectId)}";

        string IPost.ExternalUrl => ObjectId;

        DateTimeOffset IPost.PostedAt => CreatedAt;

        string? IPost.ProfileUrl => CreatedBy;

        IEnumerable<IPostThumbnail> IPost.Thumbnails => Attachments;
    }
}
